Fix alter table regression case.

Aegeaner opened the pull request:
- Feature not supported: 

Currently, GPDB doesn't support alter type on primary key and unique
constraint column. Because it requires drop - recreate logic.
The drop currently only performs on master which lead error when
recreating index (since recreate index will dispatch to segments and
there still old constraint index exists)

- Fix lockmode misuse in `RenameRelation`:

In `RenameRelation`:
https://github.com/greenplum-db/gpdb-postgres-merge/blob/ff4f020c6761a8ee1b5e79d259b401ed0a55c5b2/src/backend/commands/tablecmds.c#L4000

It used `AccessExclusiveLock` unconditionally to open the target relation, but for index statement, it should use `ShareUpdateExclusiveLock`:

https://github.com/greenplum-db/gpdb-postgres-merge/blob/ff4f020c6761a8ee1b5e79d259b401ed0a55c5b2/src/backend/commands/tablecmds.c#L3957

Before this fix, there was a diff in the test case:
```
diff -I HINT: -I CONTEXT: -I GP_IGNORE: -U3 /home/gpadmin/gpdb-postgres-merge/src/test/regress/expected/alter_table.out /home/gpadmin/gpdb-postgres-merge/src/test/regress/results/alter_table.out
--- /home/gpadmin/gpdb-postgres-merge/src/test/regress/expected/alter_table.out 2022-12-14 13:17:22.965202903 +0800
+++ /home/gpadmin/gpdb-postgres-merge/src/test/regress/results/alter_table.out  2022-12-14 13:17:23.089203851 +0800
@@ -260,9 +260,11 @@
 ORDER BY relation::regclass::text COLLATE "C";
               relation              |           mode
 ------------------------------------+--------------------------
+ alter_idx_rename_test_idx_2        | AccessExclusiveLock
  alter_idx_rename_test_idx_2        | ShareUpdateExclusiveLock
+ alter_idx_rename_test_parted_idx_2 | AccessExclusiveLock
  alter_idx_rename_test_parted_idx_2 | ShareUpdateExclusiveLock
-(2 rows)
+(4 rows) COMMIT;
 BEGIN;
```

- Different behaviours in `AlterTableGetLockLevel / AlterTableGetRelOptionsLockLevel`:

In case `AT_ResetRelOptions`, the lock mode is determined in `AlterTableGetRelOptionsLockLevel()`, Setting multiple options together uses the highest lock level required for any option. And in GPDB, the highest lockmode is set by this relOption:

But `allocate_reloption()` hard-coded the lockmode as `AccessExclusiveLock`:
https://github.com/greenplum-db/gpdb-postgres-merge/blob/e1e395db886cdfc0fd19c89aa4f27cdbdc09cd28/src/backend/access/common/reloptions.c#L678

This can be resolved by backporting an upstream commit: https://github.com/postgres/postgres/commit/69f94108079d70093b59096a3ae0ad82c842b4c0

michaelpq committed:
Allow definition of lock mode for custom reloptions

Relation options can define a lock mode other than AccessExclusiveMode
since 47167b7, but modules defining custom relation options did not
really have a way to enforce that.  Correct that by extending the
current API set so as modules can define a custom lock mode.

Author: Michael Paquier
Reviewed-by: Kuntal Ghosh
Discussion: https://postgr.es/m/20190920013831.GD1844@paquier.xyz

kainwen commented on src/backend/commands/tablecmds.c:
```diff
@@ -3948,9 +3948,9 @@ RenameRelation(RenameStmt *stmt)
 	 * escalation.  However, because ALTER INDEX can be used with any relation
 	 * type, we mustn't believe without verification.
 	 */
+	LOCKMODE	lockmode;
```
Why move this declaration out?

Aegeaner replied:
To reuse the same lockmode for index statement at line 4000.

kainwen commented on src/backend/commands/tablecmds.c:
```diff
@@ -3997,7 +3997,7 @@ RenameRelation(RenameStmt *stmt)
 	if (gp_allow_rename_relation_without_lock)
 		targetrelation = fake_relation_open(relid);
 	else
-		targetrelation = relation_open(relid, AccessExclusiveLock);
```
Will main branch has the same issue here?

Aegeaner replied:
Yes, it also has this issue.

kainwen replied:
> Yes, it also has this issue.

@Aegeaner So let's first open a PR to main branch and after main branch pushed then cherry-pick to merge branch.

Plus, what issue for this issue? why there is no change of case?

Aegeaner replied:
This is because in the GPDB-specific code in `RenameRelation`, it used `AccessExclusiveLock` unconditionally to open the target relation, but for index statement, it should use `ShareUpdateExclusiveLock`. After this fix, it aligned with upstream, so there is no change in test case.

Before this change, there was a diff in the test case:
```
diff -I HINT: -I CONTEXT: -I GP_IGNORE: -U3 /home/gpadmin/gpdb-postgres-merge/src/test/regress/expected/alter_table.out /home/gpadmin/gpdb-postgres-merge/src/test/regress/results/alter_table.out
--- /home/gpadmin/gpdb-postgres-merge/src/test/regress/expected/alter_table.out 2022-12-14 13:17:22.965202903 +0800
+++ /home/gpadmin/gpdb-postgres-merge/src/test/regress/results/alter_table.out  2022-12-14 13:17:23.089203851 +0800
@@ -260,9 +260,11 @@
 ORDER BY relation::regclass::text COLLATE "C";
               relation              |           mode
 ------------------------------------+--------------------------
+ alter_idx_rename_test_idx_2        | AccessExclusiveLock
  alter_idx_rename_test_idx_2        | ShareUpdateExclusiveLock
+ alter_idx_rename_test_parted_idx_2 | AccessExclusiveLock
  alter_idx_rename_test_parted_idx_2 | ShareUpdateExclusiveLock
-(2 rows)
+(4 rows) COMMIT;
 BEGIN;
```

kainwen replied:
Let's open a PR to main branch for this commit and let others review, thanks.

Aegeaner replied:
Should I copy the related test case (See above) to the main branch PR too?

Aegeaner replied:
@kainwen Perhaps we don't need to open a PR to main branch, since this lock relaxing for renaming index statements improvement was only introduced in PG 12, by the following two commits:

[Lower lock level for renaming indexes ](https://github.com/greenplum-db/gpdb-postgres-merge/commit/1b5d797cd4f7133ff0d18e123fcf41c67a5a7b0b)

[Ensure correct lock level is used in ALTER ... RENAME ](
https://github.com/greenplum-db/gpdb-postgres-merge/commit/6c8d1c197bd67e78c12c68d8f0bd111a59e7bccd)

Aegeaner replied:
`RelationReloadIndexInfo()` has lowered the lock level needed to reload relcache for index relations to `AccessShareLock`, which provided the possibility for relaxing lock level for index statements, but for update correctness, it still used `ShareUpdateExclusiveLock` lockmode for indexes renaming statements.

adam8157 replied:
The main branch has the [Lower lock level for renaming indexes](https://github.com/greenplum-db/gpdb-postgres-merge/commit/1b5d797cd4f7133ff0d18e123fcf41c67a5a7b0b) and it grabs an `AccessExclusiveLock` for `GpRenameChildPartitions()`.

I think the main branch needs a PR.

kainwen commented on src/test/regress/expected/alter_table.out:
```diff
@@ -1081,6 +1081,8 @@ drop table atacc1;
 alter table atacc1 add constraint atacc1_fkey foreign key (a) references atacc1 (a) not valid;
 alter table atacc1 validate constraint atacc1_fkey, alter a type bigint;
+ERROR:  cannot alter column with primary key or unique constraint
+HINT:  DROP the constraint first, and recreate it after the ALTER
```
This seems related to the feature not support. Looks OK. Do we have documented this?

Aegeaner replied:
yes it was documented as ["Note that you cannot alter column data types that are being used as distribution or partitioning keys. "](https://docs.vmware.com/en/VMware-Tanzu-Greenplum/6/greenplum-database/GUID-ref_guide-sql_commands-ALTER_TABLE.html)

kainwen commented on src/test/regress/expected/alter_table.out:
```diff
@@ -2781,19 +2783,19 @@ select * from my_locks order by 1;
------------+--------------------------
- alterlock | ShareUpdateExclusiveLock
- pg_toast  | ShareUpdateExclusiveLock
+  relname  |    max_lockmode     
```
I cannot match the test case with c code change. Is it just a test change? Can you explain a bit why the test change?

Aegeaner replied:
Yes, this is a different behaviour between PG and GP. It's described in the main PR body:

- Different behaviours in `AlterTableGetLockLevel / AlterTableGetRelOptionsLockLevel`:

In case `AT_ResetRelOptions`, the lock mode is determined in `AlterTableGetRelOptionsLockLevel()`, Setting multiple options together uses the highest lock level required for any option. And in GPDB, the highest lockmode is set by this relOption:

```
(gdb) p *relOpts[i]
$7 = {name = 0xe946ef "autosummarize", desc = 0xe9b82a "Enables automatic summarization on this BRIN index", kinds = 1024, lockmode = 8, namelen = 13, type = RELOPT_TYPE_BOOL}
```

which is AccessExclusiveLock.

kainwen replied:
@Aegeaner from the commit message, it just states why the difference.

To the problem itself, what lock mode should it hold and why?

Aegeaner replied:
It should be `AccessExclusiveLock` as it currently is, as there is a pre-defined relOption for `autosummarize`:
https://github.com/greenplum-db/gpdb-postgres-merge/blob/iteration_REL_12_12/src/backend/access/common/reloptions.c#L106~L111

```
{
			"autosummarize",
			"Enables automatic summarization on this BRIN index",
			RELOPT_KIND_BRIN,
			AccessExclusiveLock
		},
```

The details are in commit: https://github.com/greenplum-db/gpdb-postgres-merge/commit/7526e10224f0792201e99631567bbe44492bbde4

Aegeaner replied:
From the comments in `src/backend/access/common/reloptions.c`:

https://github.com/greenplum-db/gpdb-postgres-merge/blob/813dec566513d76a16d9e27d0bf06cfeb1d4c0e7/src/backend/access/common/reloptions.c#L62

> The default choice for any new option should be AccessExclusiveLock.
> In some cases the lock level can be reduced from there, but the lock
> level chosen should always conflict with itself to ensure that multiple
> changes aren't lost when we attempt concurrent changes.
> The choice of lock level depends completely upon how that parameter
> is used within the server, not upon how and when you'd like to change it.
> Safety first. Existing choices are documented here, and elsewhere in
> backend code where the parameters are used.
> 
> In general, anything that affects the results obtained from a SELECT must be
> protected by AccessExclusiveLock.

This is also aligned with upstream.

kainwen replied:
I test this under Postgres, Postgres use `ShareUpdateExclusiveLock`. I still do not understand why Greenplum has to use `AccessExclusiveLock`.

Aegeaner replied:
@kainwen Sorry the previous explanation was wrong, the lockmode was updated by the `fillfactor` option:
```
(gdb) p *relOpts[i]
$4 = {name = 0x23bed70 "fillfactor", desc = 0x23bed98 "Packs bitmap index pages only to this percentage", kinds = 4096, lockmode = 8, namelen = 10,
  type = RELOPT_TYPE_INT}
```
This is a GP-only reloption and the default value is `ShareUpdateExclusiveLock`, there should have been a lock upgrading happened somewhere:
https://github.com/greenplum-db/gpdb-postgres-merge/blob/d789f7d64aea5863d04be0805e7d9cfb94941a53/src/backend/access/common/reloptions_gp.c#L68

kainwen commented on src/test/regress/expected/alter_table.out:
```diff
@@ -4186,7 +4188,7 @@ DROP TABLE fail_part;
 CREATE VIEW at_v1 AS SELECT 1 as a;
 ALTER TABLE at_v1 ATTACH PARTITION dummy default;
-ERROR:  "at_v1" is not a partitioned table or index
+ERROR:  "at_v1" is not a partitioned table or index (parse_utilcmd.c:5075)
```
Can you use initfile to filter out `parse_utilcmd.c:5075`? Otherwise, each time modification of the file, might lead to the case failure.

Aegeaner replied:
Good idea.

Aegeaner replied:
Fixed in this commit: ec034aef286ef14a4d43e773d318bcac6e8626b3

kainwen reviewed (commented):
Mixing 3 parts in a commit is really hard for the reviewer to understand.  I cannot match each part. If it is an easy job, can you split it into commits?

Aegeaner commented:
I will split this PR into commits later.

Aegeaner committed:
Fix Lockmode misiuse in RenameRelation.

In RenameRelation, it used AccessExclusiveLock unconditionally to open the target relation, but for index statement, it should use ShareUpdateExclusiveLock.

Aegeaner committed:
Fix results of features not supported in alter_table test case.

Currently, GPDB doesn't support alter type on primary key and unique
constraint column. Because it requires drop - recreate logic.
The drop currently only performs on master which lead error when
recreating index (since recreate index will dispatch to segments and
there still old constraint index exists)

Aegeaner committed:
Add regex to test init file to filter out line number in error msg in parse_utilcmd.c
